File: frontend/src/api/conditions.ts
import { apiClient } from './client'

export interface ConditionsPayload {
  experiment_fk: number
  experiment_id: string
  experiment_type?: string
  temperature_c?: number
  initial_ph?: number
  rock_mass_g?: number
  water_volume_mL?: number
  particle_size?: string
  feedstock?: string
  reactor_number?: number
  stir_speed_rpm?: number
  initial_conductivity_mS_cm?: number
  room_temp_pressure_psi?: number
  rxn_temp_pressure_psi?: number
  co2_partial_pressure_MPa?: number
  core_height_cm?: number
  core_width_cm?: number
  confining_pressure?: number
  pore_pressure?: number
}

export interface ConditionsResponse extends ConditionsPayload {
  id: number
  water_to_rock_ratio?: number | null
  created_at: string
}

export const conditionsApi = {
  create: (payload: ConditionsPayload) =>
    apiClient.post<ConditionsResponse>('/conditions', payload).then((r) => r.data),

  getByExperiment: (experimentId: string) =>
    apiClient.get<ConditionsResponse>(`/conditions/by-experiment/${experimentId}`).then((r) => r.data),

  patch: (conditionsId: number, payload: Partial<ConditionsPayload>) =>
    apiClient.patch<ConditionsResponse>(`/conditions/${conditionsId}`, payload).then((r) => r.data),
}

File: backend/api/schemas/conditions.py
from __future__ import annotations
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict


class ConditionsCreate(BaseModel):
    experiment_fk: int
    experiment_id: str
    temperature_c: Optional[float] = None
    initial_ph: Optional[float] = None
    rock_mass_g: Optional[float] = None
    water_volume_mL: Optional[float] = None
    experiment_type: Optional[str] = None
    reactor_number: Optional[int] = None
    feedstock: Optional[str] = None
    particle_size: Optional[str] = None
    stir_speed_rpm: Optional[float] = None
    room_temp_pressure_psi: Optional[float] = None
    rxn_temp_pressure_psi: Optional[float] = None
    initial_conductivity_mS_cm: Optional[float] = None
    co2_partial_pressure_MPa: Optional[float] = None
    confining_pressure: Optional[float] = None
    pore_pressure: Optional[float] = None
    flow_rate: Optional[float] = None
    initial_nitrate_concentration: Optional[float] = None
    initial_dissolved_oxygen: Optional[float] = None
    initial_alkalinity: Optional[float] = None
    total_ferrous_iron: Optional[float] = None
    core_height_cm: Optional[float] = None
    core_width_cm: Optional[float] = None
    core_volume_cm3: Optional[float] = None


class ConditionsUpdate(BaseModel):
    temperature_c: Optional[float] = None
    initial_ph: Optional[float] = None
    rock_mass_g: Optional[float] = None
    water_volume_mL: Optional[float] = None
    experiment_type: Optional[str] = None
    reactor_number: Optional[int] = None
    feedstock: Optional[str] = None
    particle_size: Optional[str] = None
    stir_speed_rpm: Optional[float] = None
    room_temp_pressure_psi: Optional[float] = None
    rxn_temp_pressure_psi: Optional[float] = None
    initial_conductivity_mS_cm: Optional[float] = None
    co2_partial_pressure_MPa: Optional[float] = None
    initial_nitrate_concentration: Optional[float] = None
    initial_dissolved_oxygen: Optional[float] = None
    initial_alkalinity: Optional[float] = None
    total_ferrous_iron: Optional[float] = None
    confining_pressure: Optional[float] = None
    pore_pressure: Optional[float] = None
    core_height_cm: Optional[float] = None
    core_width_cm: Optional[float] = None
    core_volume_cm3: Optional[float] = None


class ConditionsResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    experiment_fk: int
    experiment_id: str
    temperature_c: Optional[float] = None
    initial_ph: Optional[float] = None
    rock_mass_g: Optional[float] = None
    water_volume_mL: Optional[float] = None
    water_to_rock_ratio: Optional[float] = None
    experiment_type: Optional[str] = None
    reactor_number: Optional[int] = None
    feedstock: Optional[str] = None
    particle_size: Optional[str] = None
    stir_speed_rpm: Optional[float] = None
    room_temp_pressure_psi: Optional[float] = None
    rxn_temp_pressure_psi: Optional[float] = None
    initial_conductivity_mS_cm: Optional[float] = None
    co2_partial_pressure_MPa: Optional[float] = None
    initial_nitrate_concentration: Optional[float] = None
    initial_dissolved_oxygen: Optional[float] = None
    initial_alkalinity: Optional[float] = None
    total_ferrous_iron: Optional[float] = None
    confining_pressure: Optional[float] = None
    pore_pressure: Optional[float] = None
    core_height_cm: Optional[float] = None
    core_width_cm: Optional[float] = None
    core_volume_cm3: Optional[float] = None
    created_at: datetime
